cdap-data-fabric/src/main/java/co/cask/cdap/spi/metadata/dataset/DatasetMetadataStorage.java

```diff
@@ -0,0 +1,432 @@
+        } else {
+          entityScopes = EnumSet.allOf(EntityScope.class);
+        }
+      } else {
+        entityScopes = Collections.singleton(EntityScope.USER);
+      }
+    }
+    MetadataSearchResponse response = searchHelper.search(new co.cask.cdap.data2.metadata.dataset.SearchRequest(
+      namespace,
+      request.getQuery(),
+      request.getTypes() == null ? Collections.emptySet() :
+        request.getTypes().stream().map(EntityTypeSimpleName::valueOf).collect(Collectors.toSet()),
+      request.getSorting() == null ? SortInfo.DEFAULT :
+        new SortInfo(request.getSorting().getKey(), SortInfo.SortOrder.valueOf(request.getSorting().getOrder().name())),
+      request.getOffset(),
+      request.getLimit(),
+      request.isCursorRequested() ? 1 : 0,
+      request.getCursor(),
+      request.isShowHidden(),
+      entityScopes
+    ));
+    // translate results back
+    List<MetadataRecord> results = new ArrayList<>(response.getResults().size());
+    response.getResults().forEach(record -> {
+      Metadata metadata = null;
+      for (Map.Entry<MetadataScope, co.cask.cdap.api.metadata.Metadata> entry : record.getMetadata().entrySet()) {
+        Metadata toAdd = new Metadata(entry.getKey(), entry.getValue().getTags(), entry.getValue().getProperties());
+        metadata = metadata == null ? toAdd : union(metadata, toAdd);
+      }
+      results.add(new MetadataRecord(record.getMetadataEntity(), metadata));
+    });
+    String cursor =
+      response.getCursors() == null || response.getCursors().isEmpty() ? null : response.getCursors().get(0);
+    return new SearchResponse(request, cursor, response.getTotal(), results);
+  }
+
+  private MetadataChange combineChanges(MetadataEntity entity,
+                                        MetadataDataset.Change userChange,
+                                        MetadataDataset.Change sysChange) {
+
+    Metadata userBefore = new Metadata(USER, userChange.getExisting().getTags(),
+                                       userChange.getExisting().getProperties());
+    Metadata sysBefore = new Metadata(SYSTEM, sysChange.getExisting().getTags(),
+                                      sysChange.getExisting().getProperties());
+    Metadata before = union(userBefore, sysBefore);
+
+    Metadata userAfter = new Metadata(USER, userChange.getLatest().getTags(), userChange.getLatest().getProperties());
+    Metadata sysAfter = new Metadata(SYSTEM, sysChange.getLatest().getTags(), sysChange.getLatest().getProperties());
+    Metadata after = union(userAfter, sysAfter);
+
+    return new MetadataChange(entity, before, after);
+  }
+
+  private static Metadata union(Metadata meta, Metadata other) {
+    return new Metadata(Sets.union(meta.getTags(), other.getTags()),
+                        ImmutableMap.<ScopedName, String>builder()
+                          .putAll(meta.getProperties())
+                          .putAll(other.getProperties()).build());
```

Comment:
This would throw an exception on key conflicts, so it's only called when one metadata is for system and the other is for user? If so it would be more clear to add a comment or name them differently or something.

Reply:
good point, you understood it exactly right.